readme.md
# Queue Shedule

[![NPM version](https://img.shields.io/npm/v/queue-schedule.svg?style=flat-square)](https://www.npmjs.com/package/queue-schedule)
[![Build status](https://travis-ci.org/e174596549/queue-schedule.svg?branch=master)](https://travis-ci.org/e174596549/queue-schedule)

Kafka is a high avaliable message queue, but it lacks of consuming message with a slow speed. Some of task with no need to finish it at none, and we want to complete it with a small cost. This is just the reason why we develop `Queue Shedule`.

## Install

```npm install queue-schedule```

## How to use

A basic example is showed as follows:

```javascript
const kafka = require('kafka-node');
const {expect} = require('chai');
const {KafkaProducer,KafkaConsumer} = require('queue-schedule');
const KAFKA_HOST = process.env.KAFKA_PEERS;
const FIST_DATA = {a:1,b:2};
const SCHEDULE_NAME1 = 'schedule1';
const TOPIC_NAME1 = 'topic.1';
const PARTITION1 = 0;

let hasDone = false;
new KafkaConsumer({
    name: 'kafka',
    topics:[TOPIC_NAME1],
    consumerOption: {
        kafkaHost: KAFKA_HOST,
        fromOffset: 'earliest',
        fetchMaxBytes: 1024*1024,
    },
    doTask:function(messages,callback) {console.log(messages);
        if (!hasDone) {
            const value = messages[0].value;
            let data = null;
            try {
                data = JSON.parse(value);
            } catch (e) {
                hasDone = true;
                console.error('parse message error',e);
                return;
            }
            expect(data).to.have.property('a').and.equal(1);
            console.log('recieve data',data);
            hasDone = true;
        }
        callback();
    },
    readCount : 1,
    pauseTime : 500,
    idleCheckInter: 10 * 1000
}).on(KafkaConsumer.EVENT_CONSUMER_ERROR,function(err) {
    console.error('consumer error',err);
    hasDone = true;
    
});


new KafkaProducer({
    name : SCHEDULE_NAME1,
    topic: TOPIC_NAME1,
    kafkarHost:KAFKA_HOST
}).addData(FIST_DATA,{},function(err) {
    if (err) {
        console.error('write to queue error',err);
        return ;
    }
    console.info('write to kafka finished');
});
```

## API

For detail usage, see the document online [here](https://yunnysunny.github.io/queue-schedule).

## License

[MIT](LICENSE)


CHANGELOG.md
# v1.1.0
## Add
1. Add the object of `globalEvent` to emit the events of client's and producer's status.
2. Use a producer pool to cache the producer with same kafka host, in other words, you will only create a single producer instance, when you instantiate diffrent `KafkaProducer` with the same kafka host.

## Remove
1. Remove the event of `EVENT_CLIENT_READY` `EVENT_CLIENT_ERROR` `EVENT_CLIENT_CLOSE` `EVENT_PRODUCER_READY` `EVENT_PRODUCER_ERROR` from `KafkaProducer`, these events mentioned above, are move to the object `globalEvent`.

# v1.0.0
## Add
1. Emit event when kafkaClient closed.
2. Add the parameter of `prepareMiddleware` to `KafkaProducer`.

## Breaking Changes
1. Remove the option of `zookeeperHost` form `KafkaProducer` and `KafkaConsumer`.
2. Remove the `manager` object.
3. Remove the function of `addDataDelay` form `KafkaProducer`.
4. Remove the parameter of `topicList` from `KafkaProducer`.
5. The function `addData` of `KafkaProducer` returns a boolean value, which indicate whether the data to send is vaild.

# v0.6.0
## Add
1. Bump the package `node-kafka` to 3.0.1.

# v0.5.3
## Fix
1. Resovle the issue of double stringify.

# v0.5.2
## Add
1. Add parameter of `fromOffset` to KafkaConsumer.

# v0.5.1
## Add
1. Add feature of emitting event of `KafkaProducer.EVENT_SEND_ERROR`.

# v0.5.0
## Add
1. Use consumerGroup instead of consumer.
2. Add the feature of using broker url to connect to the kafka client.

# v0.3.1
## Add
1. Use HighLevelProducer instead of Producer.
2. Add the feature of setting the groupId of consumer.
3. Add the parameter of the count of messages published to the event of `EVENT_DELAY_MESSAGE_SEND_FINISHED`

# v0.2.1
## Add
1. Do code refactoring, create kafka client in the KafkaProducer and KafkaConsumer.
2. Add feature of emitting event of `KafkaProducer.EVENT_CLIENT_READY` `KafkaProducer.EVENT_CLIENT_ERROR` `KafkaProducer.EVENT_PRODUCER_READY` `KafkaProducer.EVENT_PRODUCER_ERROR` `KafkaConsumer.EVENT_CLIENT_READY` `KafkaConsumer.EVENT_CLIENT_ERROR` `KafkaConsumer.EVENT_CONSUMER_ERROR`.
## Fix
1. Fxed the error of breaking down when call addKafkaSchedule twice with the same name.

# v0.2.0
## Add
1. Add some testcases.
2. Add the feature of emitting event of  `KafkaProducer.EVENT_DELAY_MESSAGE_SEND_FINISHED` when delay message publish finished.

# v0.1.0
## Add
1. Add feature f sending message delay.
2. Add travis-ci.yml.



test/race.js
const async = require('async');
async.race([
    function(callback) {
        setTimeout(function() {
            console.log('200');
            callback(null, 'one');
        }, 200);
    },
    function(callback) {
        setTimeout(function() {
            console.log('100');
            callback(null, 'two');
        }, 100);
    }
],
// main callback
function(err, result) {
    // the result will be equal to 'two' as it finishes earlier
    console.log(result);
});

test/consumer.js
var kafka = require('kafka-node');

var topic = 'first';


var consumer = //new Consumer(client, topics, options);
new kafka.ConsumerGroup({
  groupId:'testbygroup',//the consumer group can only be found in kafka manange, not in zookeeper
  kafkaHost:process.env.KAFKA_PEERS,
  autoCommit: true,
  fromOffset:'earliest'
},[topic]);
// var offset = new Offset(client);

consumer.on('message', function (message) {
  console.log(message);
});

consumer.on('error', function (err) {
  console.log('error', err);
});



test/producer.js
var kafka = require('kafka-node');
var Producer = kafka.Producer;
var KeyedMessage = kafka.KeyedMessage;
var Client = kafka.Client;
var client = new Client(process.env.ZOOKEEPER_PEERS);
//var argv = require('optimist').argv;
var topic = 'first';
var p = 0;
var a = 0;
var producer = new Producer(client, { requireAcks: 1 });

producer.on('ready', function () {
  var message = new Date().toLocaleString();
  var keyedMessage = new KeyedMessage('keyed', 'a keyed message');

  producer.send([
    { topic: topic, partition: p, messages: [message, keyedMessage], attributes: a }
  ], function (err, result) {
    if (err) {
      console.error(err);
    } else {
      console.log(message,result);
    }
    
    process.exit();
  });
});

producer.on('error', function (err) {
  console.log('error', err);
});

lib/manager.js
const KafkaProducer = require('./KafkaProducer');
// const kafka = require('kafka-node');
const scheduleManager = new Map();
const SCHEDULE_STATUS = {
    PENDING : 1,
    SUCCESS : 2,
    FAILED : 3
};

class ScheduleItem {
    /**
     * Creates an instance of ScheduleItem.
     * @param {KafkaProducer} data 
     * @param {Number} status 
     * @param {Promise} promise 
     * @memberof ScheduleItem
     */
    constructor(data, status, promise) {
        this.data = data;
        this.status = status;
        this.promise = promise;
    }
}
/**
 * 
 * @param {ProducerOption} options 
 */
const _genNewProducer = function(options,taskData,callback) {
    if (options.host) {
        options.zookeeperHost = options.host;
    }
    const kafkaProducer = new KafkaProducer(options);
    const item = new ScheduleItem(kafkaProducer,SCHEDULE_STATUS.PENDING);
    item.promise = new Promise(function(resolve,reject) {
        kafkaProducer.on(KafkaProducer.EVENT_PRODUCER_READY,function(err) {
            if (err) {
                item.status = SCHEDULE_STATUS.FAILED;
                callback(err);
                return reject(err);
            }
            item.status = SCHEDULE_STATUS.SUCCESS;
            if (taskData) {
                kafkaProducer.addData(taskData);
            }
            callback();
            resolve();
        });
    });
    scheduleManager.set(options.name,item);
};

/**
 * @callback AddScheduleCallback
 * @param {Error} err
 */

/**
 * Add a schedule to the manager, it will return a new instance if none exist, otherwise will return the exist one.
 * **Attention:** Make sure you give a unique name for diffrent proudcers, otherwise it will act as a same one.
 *   
 * @param  {ProducerOption} options the option
 * @param  {Object=} [taskData=null] The data of task, send to kafka after created.
 * @param  {AddScheduleCallback=} [callback=function(){}] Called when the internal kafka producer is ready or error.
 */
exports.addKafkaSchedule = function(options, taskData = null, callback = function() {}) {
    const name = options.name;
    const scheduleNow = scheduleManager.get(name);
    if (scheduleNow) {
        if (scheduleNow.status === SCHEDULE_STATUS.SUCCESS) {
            if (taskData) {
                scheduleNow.data.addData(taskData);
            }
            return callback();
        }
        if (scheduleNow.status === SCHEDULE_STATUS.PENDING) {
            if (!taskData) {
                return callback();
            }
            return scheduleNow.promise.then(function() {
                callback();
                scheduleNow.data.addData(taskData);
            },function(err) {
                callback(err);
            });
        }
        //failed before, re-try
        _genNewProducer(options,taskData,callback);
    } else {

        _genNewProducer(options,taskData,callback);
    }
};
/**
 * Get the producer object by the name of a schedule.
 * 
 * @param {String} name
 * @return {KafkaProducer}
 */
exports.getProducerByScheduleName = function(name) {
    const schedule = scheduleManager.get(name);
    if (!schedule) {
        return null;
    }
    return schedule.data;
};


lib/KafkaProducer.js
const EventEmitter = require('events');
// const kafka = require('kafka-node');
const {getProducer} = require('./producer_pool');
/**
 * @function PrepareMiddleware
 * 
 * @param {Object} data the data want to send
 * @returns {Boolean} whether send to kafka server
 */

 /**
  * @typedef {Object} SendOption
  * 
  * @property {String} [topicSuffix=''] The suffix of the topic name.
  */

 /**
  * @typedef {Object} ProducerOption
  * 
  * @class KafkaProducer
  * @param {String} name The name of current instance.
  * @param {String=} kafkaHost The host of the broker of kafka, when both of `zookeeperHost` and `kafkaHost` passed, the `kafkaHost` has higher priority.
  * @param {String=} topic The topic where you save data in it.
  * @param {Number=} delayInterval When pass this parameter, messages will publish to kafka every `option.delayInterval` ms, otherwise messages will publish to kafka at once.
  * @param {PrepareMiddleware=} prepareMiddleware
  */

/**
 * The class of the producer of Kafka
 * @class KafkaProducer
 * @extends {EventEmitter}
 */
class KafkaProducer extends EventEmitter {
    /**
     * Creates an instance of KafkaProducer.
     * 
     * @param {ProducerOption} option
     * @memberof KafkaProducer
     */
    constructor({
        name,
        topic,
        kafkaHost,
        delayInterval,
        prepareMiddleware
    }) {
        super();
        this.name = name;
        this.producer = null;
        this._clientPromise = null;
        this._clientOk = false;
        if (!kafkaHost) {
            throw new Error('You must giva a  kafkaHost.');
        }
        this.topic = topic;
        if (!this.topic) {
            throw new Error('You must give the topic paramemter.');
        }
        this._createProducer(kafkaHost);

        this.delayInterval = delayInterval;
        this._delayData = {};
        this._prepareMiddleware = null;
        if (typeof (prepareMiddleware) === 'function') {
            this._prepareMiddleware = prepareMiddleware;
        }
        this._delayTimer = null;
        if (delayInterval > 0) {
            this._doSendDataTimer();
        }
    }
    /**
     * @private
     * 
     * @memberof KafkaProducer
     */
    _cancleTimer() {
        if (this._delayTimer) {
            clearTimeout(this._delayTimer);
        }
    }
    /**
     * @private
     * @param {String} zookeeperHost 
     * 
     * @memberof KafkaProducer
     */

    _createProducer(kafkaHost) {
        // const client = new kafka.KafkaClient({kafkaHost: kafkaHost});
        const _this = this;

        _this._clientPromise = new Promise(function(resolve,reject) {

            const producerCreater = getProducer(kafkaHost);
            if (producerCreater.producer) {
                _this.producer = producerCreater.producer;
                _this._clientOk = true;
                resolve();
                return;
            }
            producerCreater.initPromise.then(function() {
                _this.producer = producerCreater.producer;
                _this._clientOk = true;
                resolve();
            }).catch(function(err) {
                _this._clientOk = false;
                reject(err);
            });
            
        }).catch(function() {
            
        });
    }

    /**
     * @private
     * 
     * @memberof KafkaProducer
     */
    _doSendDataTimer() {
        const _this = this;
        const _delayData = this._delayData;
        this._delayTimer = setTimeout(function delayProcess() {
            if (!_this.producer) {
                return _this._doSendDataTimer();
            }
            const keys = Object.keys(_delayData);
            const arrayLen = keys.length;
            if (arrayLen === 0) {
                return _this._doSendDataTimer();
            }
            for (var i=0;i<arrayLen;i++) {
                const topicSuffix = keys[i];
                const topicData = _delayData[topicSuffix];
                const dataArray = topicData.splice(0);
                if (dataArray.length > 0) {
                    _this.producer.send(
                        _this._createSendData(dataArray,{topicSuffix}),function(err) {
                            if (err) {
                                _this.emit(KafkaProducer.EVENT_SEND_ERROR,err,dataArray);
                            }
                            _this.emit(
                                KafkaProducer.EVENT_DELAY_MESSAGE_SEND_FINISHED,
                                err,dataArray.length
                            );
                        }
                    );
                }

            }
            _this._doSendDataTimer();
        },this.delayInterval);
        

    }
    /**
     * @private
     * 
     * @param {Object} taskData 
     * @param {SendOption} options
     * @returns 
     * @memberof KafkaProducer
     */
    _createSendData(taskData, options) {
        const suffix = options.topicSuffix || '';
        return [{
            topic: this.topic + suffix,
            messages: taskData,
        }];
    }
    _getSendData(taskData) {
        return typeof(taskData) === 'object' ?
        JSON.stringify(taskData):
        taskData;
    }
    /**
     * 
     * @private
     * @param {Object} taskData 
     * @param {SendOption} options
     * @param {Function} callback 
     * @memberof KafkaProducer
     */
    _sendAtOnce(taskData,options,callback) {
        const _this = this;
        this.producer.send(
            this._createSendData(this._getSendData(taskData), options),
            function(err,data) {
                callback(err,data);
                if (err) {
                    _this.emit(KafkaProducer.EVENT_SEND_ERROR,err,taskData);
                }
            }
        );
    }

    /**
     * Send data to kafka, it will send the data to kafka every `delayInterval` ms when `delayInterval` is set. It will wait the client i
     * 
     * @param {Object} taskData 
     * @param {SendOption} options
     * @param {Function=} [callback=function(err) {}] 
     * @returns {Boolean} Whether the taskData is valid.
     * @memberof KafkaProducer
     */
    addData(taskData, options = { topicSuffix : '' }, callback = function() {}) {
        if (this._prepareMiddleware && this._prepareMiddleware(taskData) === false) {
            return false;
        }
        const topicSuffix = options.topicSuffix || '';
        if (this.delayInterval > 0) {
            if (!this._delayData[topicSuffix]) {
                this._delayData[topicSuffix] = [];
            }
            this._delayData[topicSuffix].push(this._getSendData(taskData));
            return true;
        }
        if (this._clientOk) {
            this._sendAtOnce(taskData,options,callback);
        } else {
            const _this = this;
            this._clientPromise.then(function() {
                _this._sendAtOnce(taskData,options,callback)
            }).catch(function(err) {
                callback(err);
            });
        }
        
        return true;
    }

}
/**
 * The event to notify that the client is ready.
 */
// KafkaProducer.EVENT_CLIENT_READY = 'eventClientReady';
/**
 * The event to notify that the client is error.
 */
// KafkaProducer.EVENT_CLIENT_ERROR = 'eventClientError';

/**
 * The event of notify that the client is closed.
 */
// KafkaProducer.EVENT_CLIENT_CLOSE = 'eventClientClose';

/**
 * The event to notify that a batch of messages have been sent finished.
 */
KafkaProducer.EVENT_DELAY_MESSAGE_SEND_FINISHED = 'eventDelayMessageSendFinished';
/**
 * The event to notify that the producer is ready.
 */
// KafkaProducer.EVENT_PRODUCER_READY = 'eventProducerReady';
/**
 * The event to notify the producer is error.
 */
// KafkaProducer.EVENT_PRODUCER_ERROR = 'eventProducerError';

/**
 * The event emitted when an error occurs after sending data to kafka.
 */
KafkaProducer.EVENT_SEND_ERROR = 'eventSendError';

module.exports = KafkaProducer;


lib/producer_pool.js
const EventEmitter = require('events');
const kafka = require('kafka-node');

const producerPool = new Map();
class GlobalProducerEvent extends EventEmitter {};

const globalEvent = new GlobalProducerEvent();
/**
 * The event to notify that the client is ready.
 */
const EVENT_CLIENT_READY = globalEvent.EVENT_CLIENT_READY = 'eventClientReady';
/**
 * The event to notify that the client is error.
 */
const EVENT_CLIENT_ERROR = globalEvent.EVENT_CLIENT_ERROR = 'eventClientError';

/**
 * The event of notify that the client is closed.
 */
const EVENT_CLIENT_CLOSE = globalEvent.EVENT_CLIENT_CLOSE = 'eventClientClose';

/**
 * The event to notify that the producer is ready.
 */
const EVENT_PRODUCER_READY = globalEvent.EVENT_PRODUCER_READY = 'eventProducerReady';

/**
 * The event to notify the producer is error.
 */
const EVENT_PRODUCER_ERROR = globalEvent.EVENT_PRODUCER_ERROR = 'eventProducerError';

exports.globalEvent = globalEvent;
/**
 * The class to init kafka producer.
 * 
 * @class ProducerCreater
 * @extends {EventEmitter}
 * @ignore
 */
class ProducerCreater extends EventEmitter {
    constructor(kafkaHost) {
        super();
        this.producer = null;
        this._init(kafkaHost);
    }
    _init(kafkaHost) {
        const client = new kafka.KafkaClient({kafkaHost: kafkaHost});
        const _this = this;
        this.initPromise = new Promise(function(resolve,reject) {
            client.on('ready',function() {
                globalEvent.emit(EVENT_CLIENT_READY,kafkaHost);
                // _this.emit(EVENT_CLIENT_READY);
            });
            client.on('error',function(err) {
                // _this._cancleTimer();
                // _this._clientOk = false;
                reject(err);
                globalEvent.emit(EVENT_CLIENT_ERROR,kafkaHost,err);
                // return ;_this.emit(EVENT_CLIENT_ERROR,err);
            });
            client.on('close',function() {
                globalEvent.emit(EVENT_CLIENT_CLOSE,kafkaHost);
                reject(new Error('The client '+ kafkaHost + ' close'));
            });
            const producer =  new kafka.HighLevelProducer(client);
            producer.on('ready', function producerReady() {
    
                _this.producer = producer;
                // _this._clientOk = true;
                resolve();
                globalEvent.emit(EVENT_PRODUCER_READY,kafkaHost);
                // return _this.emit(EVENT_PRODUCER_READY);
                
            }); 
            producer.on('error',function producerError(err) {
                // _this._cancleTimer();
                // _this._clientOk = false;
                reject(err);
                globalEvent.emit(EVENT_PRODUCER_ERROR,kafkaHost,err);
                // return _this.emit(EVENT_PRODUCER_ERROR,err);
            });
        });
        
    }
}


exports.ProducerCreater = ProducerCreater;

/**
 * Get an instance of producer.
 * 
 * @ignore
 * @param {String} kafkaHost
 * @returns {ProducerCreater}
 */
exports.getProducer = function(kafkaHost) {
    const clientOld = producerPool.get(kafkaHost);
    if (clientOld) {
        return clientOld;
    }
    const newProducer = new ProducerCreater(kafkaHost);
    producerPool.set(kafkaHost,newProducer);
    return newProducer;
};

lib/KafkaConsumer.js
const EventEmitter = require('events');
const kafka = require('kafka-node');
/**
 * @callback DoTask
 * 
 * @param {Object[]} messages
 */


/**
 * @typedef ConsumerOption
 * 
 * The options passed to [kafka.ConsumerGroup](https://github.com/SOHU-Co/kafka-node#consumergroup) 
 * 
 * @param {String=} kafkaHost connect directly to kafka broker (instantiates a KafkaClient), when not passed, it will use KafkaConsumerOption.kafkaHost in default.
 * @param {Boolean} [ssl=false] optional (defaults to false) or tls options hash.
 * @param {String=} [encoding='utf8'] default is utf8, use 'buffer' for binary data.
 * @param {String=} groupId it will use KafkaConsumerOption.groupId in default.
 * @param {Number=} [fetchMaxBytes=1024*1024]
 * @param {Number} [sessionTimeout=15000]
 * @param {String[]} [protocol=['roundrobin']] An array of partition assignment protocols ordered by preference. 'roundrobin' or 'range' string for built ins
 * @param {String} [fromOffset=earliest] Offsets to use for new groups other options could be 'earliest' or 'none' (none will emit an error if no offsets were saved) quivalent to Java client's auto.offset.reset
 * @param {String} [outOfRangeOffset=earliest] how to recover from OutOfRangeOffset error (where save offset is past server retention) accepts same value as fromOffset
 * @param {Boolean} [migrateHLC=false]
 * @param {Boolean} [migrateRolling=true]
 */

/**
 * @typedef KafkaConsumerOption
 * 
 * @param {String} name The name of current instance.
 * @param {String=} kafkaHost The host of the broker of kafka.
 * @param {String[]} topics The topics that will be consumed.
 * @param {ConsumerOption=} consumerOption The option to create a new instance of `Kafka.ConsumerGroup`.
 * @param {Number} readCount After reading the count of `readCount`, the consumer will be paused.
 * @param {Number} pauseTime The duration of pause time, after that the consumer will be continued.
 * @param {DoTask} doTask The consume process function.
 * @param {Number} idleCheckInter The instance of KafkaConsumer has a timer inner, to check whether the process of `doTask` is idle. The timer will trigger every `idleCheckInter` ms. 
 */

/**
 * The class of KafkaConsumer
 * @class KafkaConsumer
 * @extends {EventEmitter}
 */

class KafkaConsumer extends EventEmitter {
    
    /**
     * Creates an instance of KafkaConsumer. It will call the function of #consumer inner.
     * 
     * @param {KafkaConsumerOption} option
     * @memberof KafkaConsumer
     */
    constructor({
        name,
        kafkaHost,
        topics,
        consumerOption = {},
        doTask,
        readCount = 100,
        pauseTime = 500,
        idleCheckInter = 1000 * 10,
        
    }) {
        super();
        this.name = name;
        this.readCount = readCount;
        this.consumer = null;
        if (!topics ||  !consumerOption) {
            throw new Error('The parameters of  topics and consumerOption must be given.');
        }
        if (!kafkaHost && !consumerOption.kafkaHost) {
            throw new Error('The kafka host must be given in kafkaHost or in consumerOption.kafkaHost.');
        }
        consumerOption.groupId = consumerOption.groupId || name;
        consumerOption.kafkaHost = consumerOption.kafkaHost || kafkaHost;
        consumerOption.fromOffset =  consumerOption.fromOffset || 'earliest';
        this._init(topics,consumerOption);
        this.pauseTime = pauseTime;
        this.doTask = doTask;
        this.lastFinishedTime = 0;
        this.idleCheckInter = idleCheckInter;
        this.messages = [];
        this.consume(doTask)
    }

    _init(topics,consumerOption) {
        // const client = new kafka.Client(zookeeperHost);
        const _this = this;
        // client.on('ready',function() {
        //     _this.emit(KafkaConsumer.EVENT_CLIENT_READY);
        // });
        // client.on('error',function(err) {
        //     _this.emit(KafkaConsumer.EVENT_CLIENT_ERROR,err);
        // });
        const names = [];
        for (const topic of topics) {
            if (typeof (topic) === 'string') {
                names.push(topic);
            } else {
                names.push(topic.topic);
            }
        }
        const consumer = this.consumer = new kafka.ConsumerGroup(
            consumerOption,names
        );
        consumer.on('error',function(err) {
            _this.emit(KafkaConsumer.EVENT_CONSUMER_ERROR,err);
        });
    }

    _continue() {
        const _this = this;
        setTimeout(function resume() {
            _this.consumer.resume();
        }, this.pauseTime);
    }
    /**
     * The consume function.
     * Do not call this function manual!
     * 
     * @param {DoTask} doTask 
     * @memberof KafkaConsumer
     */
    consume(doTask) {
        let messages = this.messages;
        const consumer = this.consumer;
        const count = this.readCount;
        const _this = this;

        consumer.on('message', function(message) {
            if (message.offset % count === 0) {
                consumer.pause();
                messages.push(message);
                _this.lastFinishedTime = new Date().getTime();
                doTask(messages.splice(0), function () {
                    _this._continue();
                });
            } else {
                messages.push(message);
                _this.lastFinishedTime = new Date().getTime();
            }
        });

        setInterval(function clear() {
            const idle = new Date().getTime() - _this.lastFinishedTime;
            if ( idle > 1000 && messages.length > 0) {
                consumer.pause();
                doTask(messages.splice(0), function() {
                    _this._continue();
                });
            }
        }, this.idleCheckInter);
    }
}
/**
 * The event to notify that the client is ready.
 */
KafkaConsumer.EVENT_CLIENT_READY = 'eventClientReady';
/**
 * The event to notify that the client is error.
 */
KafkaConsumer.EVENT_CLIENT_ERROR = 'eventClientError';
/**
 * The event to notify that an error ocurred in consumer.
 */
KafkaConsumer.EVENT_CONSUMER_ERROR = 'eventConsumerError';

module.exports = KafkaConsumer;
